### PR TITLE
[MINI-5751] Universal Bridge: Communication between the Host App and MiniApps

### PR DESCRIPTION
**Pull request by shohiebsense**
# Description
Universal Bridge feature:
Any MiniApp -> the HostApp using `sendJsonToHostapp` ActionType
The HostApp -> active MiniApps using `miniappreceivejsoninfo` NativeEventType

## Links
[MINI-5751](https://jira.rakuten-it.com/jira/browse/MINI-5751)
[MINI-5755](https://jira.rakuten-it.com/jira/browse/MINI-5755)

## Video
~~Demo: https://rak.box.com/s/k2lb6ka73n0c4v7a2i95mx068ovyhwjb OBSOLETE
If it has the universal bridge message, the toolbar icon to send would show. Otherwise, it is hidden.~~

https://rak.box.com/s/d1ijrrrhuirs6pql7bvf6po0dtqivyf6

# Checklist
- [x] I have read the [contributing guidelines](https://github.com/rakutentech/android-miniapp/blob/master/.github/CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](https://github.com/rakutentech/android-miniapp/blob/master/.github/CONTRIBUTING.md#sdk-development-learning-path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `./gradlew check` without errors

In case code coverage is dropping, Jacoco doesn't go well with `lateinit` property.
https://github.com/jacoco/jacoco/issues/1182

### CODE REVIEW

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/Type.kt [hunk 1]
```diff
@@ -51,5 +52,6 @@ enum class NativeEventType(val value: String) {
     MINIAPP_ON_PAUSE("miniapppause"),
     MINIAPP_ON_RESUME("miniappresume"),
     MINIAPP_SECURE_STORAGE_READY("miniappsecurestorageready"),
-    MINIAPP_SECURE_STORAGE_LOAD_ERROR("miniappsecurestorageloaderror")
+    MINIAPP_SECURE_STORAGE_LOAD_ERROR("miniappsecurestorageloaderror"),
+    RECEIVE_UNIVERSAL_JSON_INFO("miniappreceivejsoninfo")
```
Comment (khairul-alam-licon) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/Type.kt:56]:
```suggestion
    RECEIVE_JSON_INFO("miniappreceivejsoninfo")
```

Comment (shohiebsense) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/Type.kt:56]: For maintainability reasons, I would suggest keeping it that way. However, I'm okay with the change either.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,9 @@
 ## CHANGELOG
 
 ### 5.1.0 (xxxx-xx-xx)
+**SDK**
+- **Feature:** Added `Universal Bridge` to communicate between MiniApps and the HostApp
```
Comment (rleojoseph) [inline CHANGELOG.md:5]: Please add the interface name that you are supporting here

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt [hunk 3]
```diff
@@ -474,6 +504,7 @@ internal object ErrorBridgeMessage {
     const val NO_IMPLEMENT_CUSTOM_PERMISSION =
         "The `MiniAppMessageBridge.requestCustomPermissions` $NO_IMPL"
     const val ERR_SHARE_CONTENT = "Cannot share content:"
+    const val ERR_UNIVERSAL_BRIDGE = "Cannot send the universal bridge:"
```
Comment (rleojoseph) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt:507]:
```suggestion
    const val ERR_UNIVERSAL_BRIDGE = "Please check the String and try again."
```

**Review by rleojoseph — COMMENTED** [review]
Thank you for the changes buddy

**Comment by shohiebsense** [comment]
Made some changes.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt [hunk 4]
```diff
@@ -411,6 +407,40 @@ open class MiniAppMessageBridge {
...
+        onError: (message: String) -> Unit
+    ) {
+        content?.let {
+            val contentJsonStr = it.toString()
+            if (contentJsonStr.isNotBlank()) {
+                onSuccess(contentJsonStr)
+                return
+            }
+        }
+        onError(ErrorBridgeMessage.ERR_UNIVERSAL_BRIDGE)
```
Comment (khairul-alam-licon) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt:441]: I'm not sure if we need to add a default implementation for `sendJsonToHostApp` in SDK side 🤔  if so, then we also need to match the same behaviour with iOS. Or, we can keep this interface unimplemented like `getUniqueId` and let the HostApp to implement.

cc: @rleojoseph @GondiTilak @munir-rakuten

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt [hunk 5]
```diff
@@ -474,6 +504,7 @@ internal object ErrorBridgeMessage {
     const val NO_IMPLEMENT_CUSTOM_PERMISSION =
         "The `MiniAppMessageBridge.requestCustomPermissions` $NO_IMPL"
     const val ERR_SHARE_CONTENT = "Cannot share content:"
+    const val ERR_UNIVERSAL_BRIDGE = "Please check the String and try again."
```
Comment (khairul-alam-licon) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt:507]: I think, we'll have real error message when there will be exception, so, this variable can be used as a prefix to represent the feature.

```suggestion
    const val ERR_UNIVERSAL_BRIDGE = "Cannot send json info:"
```

**Review by khairul-alam-licon — COMMENTED** [review]
Looking good 👍 
I have left few feedbacks meanwhile. Also, I'm curious to know if we need to add `sendJsonToMiniApp` interface in SDK side as well 🤔

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt (hunk 3 above)
Comment (shohiebsense) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt]: @khairul-alam-licon

**Comment by shohiebsense** [comment]
Thanks @khairul-alam-licon for the review

Ok, then the 'Universal Bridge' would lose its the way it is naming, why can't we call it  `Send /Receive Json Info` instead

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt [hunk 6]
```diff
@@ -411,6 +407,40 @@ open class MiniAppMessageBridge {
...
+                bridgeExecutor.postValue(callbackObj.id, value.toString())
+            },
+            onError = { message ->
+                bridgeExecutor.postError(
+                    callbackObj.id,
+                    message
+                )
+            }
+        )
+    }
```
Comment (munir-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt:424]: I can see a couple of things wrong here, we are not suppose return the param here, we should parse the contents in the param and send it
```
private fun onSendJsonToHostApp(callbackId: String, jsonStr: String) = try {
        val callbackObj = Gson().fromJson(jsonStr, JsonInfoCallbackObj::class.java)
        val successCallback = { info: String ->
            bridgeExecutor.postValue(callbackId, info)
        }
        val errorCallback = { message: String ->
            bridgeExecutor.postError(
                callbackId, "${ErrorBridgeMessage.ERR_UNIVERSAL_BRIDGE} $message"
            )
        }

        sendJsonToHostApp(callbackObj.param.jsonInfo.content, successCallback, errorCallback)
    } catch (e: Exception) {
        bridgeExecutor.postError(callbackId, "${ErrorBridgeMessage.ERR_UNIVERSAL_BRIDGE} ${e.message}")
    }
``` 

The object is suppose to look like this
```
@Keep
internal data class JsonInfo(val content: String)

@Keep
internal data class JsonInfoCallbackObj(val param: JsonInfoParam) {

    @Keep
    internal data class JsonInfoParam(val jsonInfo: JsonInfo)
}
```

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt (hunk 4 above)
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt:441]: There should not be a default implementation for this
```
open fun sendJsonToHostApp(
        info: String,
        onSuccess: (info: String) -> Unit,
        onError: (message: String) -> Unit
    ) {
        throw MiniAppSdkException(ErrorBridgeMessage.NO_IMPL)
    }
```

**Review by munir-rakuten — COMMENTED** [review]
Left some comments, please check it,thanks.
I can not see a demo app implementation, is it for another ticket?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt (hunk 3 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt]: Yeah, we could use `"Cannot send json info:"` instead as a prefix like the other existing features. But, `"Please check the String and try again."` is also fine.
I'm giving screenshots of both cases when sending an error message from HostApp using the interface.

Screen Shot 2022-12-01 at 16 51 43
Screen Shot 2022-12-01 at 16 53 13

In that case, you can ignore [this comment](https://github.com/rakutentech/android-miniapp/pull/524#discussion_r1036742817).

**Comment by khairul-alam-licon** [comment]
@shohiebsense Yeah, I agree with you about naming. It's confusing and hard when there are multiple scopes, specially, this feature is called **UniversalBridge** including two scopes like **Send** & **Receive**. We can actually use all keywords depending on various purposes. 
e.g. we can write in other places e.g. `NativeEventType.RECEIVE_UNIVERSAL_JSON_INFO` to define receiving scope, and `ErrorBridgeMessage.ERR_UNIVERSAL_BRIDGE` to define the full feature. Also, `ActionType.SEND_JSON_TO_HOSTAPP("sendJsonToHostapp")` to define original bridge's action.

Please feel free about naming what you feel right and consider [this comment](https://github.com/rakutentech/android-miniapp/pull/524#discussion_r1036704557) as optional.

**Comment by shohiebsense** [comment]
Though I prefer it not to be wrapped by try-catch, changes made, also some refactors, let me know as soon as possible if you don't like it.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/settings/cache/Cache.kt [hunk 7]
```diff
@@ -137,6 +142,7 @@ internal class Cache(
         private const val UNIQUE_ID_ERROR = "unique_id_error"
         private const val MESSAGING_UNIQUE_ID_ERROR = "messaging_unique_id_error"
         private const val MAUID_ERROR = "mauid_error"
+        private const val UNIVERSAL_BRIDGE_MESSAGE = "universal_bridge_error"
```
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/settings/cache/Cache.kt:145]:
```suggestion
        private const val UNIVERSAL_BRIDGE_MESSAGE = "universal_bridge_message"
```

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/fragments/MiniAppDisplayFragment.kt [hunk 8]
```diff
@@ -315,6 +320,24 @@ class MiniAppDisplayFragment : BaseFragment(), PreloadMiniAppWindow.PreloadMiniA
                     AppPermission.getDeviceRequestCode(miniAppPermissionType)
                 )
             }
+
+            override fun sendJsonToHostApp(
+                jsonStr: String,
+                onSuccess: (String) -> Unit,
```
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/fragments/MiniAppDisplayFragment.kt:326]: (optional) you also can use `onSuccess` to send a confirmation message to MiniApp. `onSuccess` will display an alert dialog like this with invoking `onSuccess(jsonStr + "has been found from MiniApp.")`

Screen Shot 2022-12-02 at 12 17 51

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/userdata/QASettingsActivity.kt [hunk 9]
```diff
@@ -117,6 +121,28 @@ class QASettingsActivity : BaseActivity() {
...
+            override fun sendJsonToHostApp(
+                jsonStr: String,
+                onSuccess: (content: String) -> Unit,
+                onError: (message: String) -> Unit
+            ) {
+                jsonStr.let {
+                    if (it.toString().isNotBlank()) {
+                        onSuccess(it)
+                        Toast.makeText(this@QASettingsActivity, it.toString(), Toast.LENGTH_LONG)
+                            .show()
```
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/userdata/QASettingsActivity.kt:140]: Do we need to put a `return` after this line? 🤔

File: testapp/src/main/res/layout/qa_settings_activity.xml [hunk 10]
```diff
@@ -247,6 +247,47 @@
...
+                app:endIconMode="clear_text">
+
+                <androidx.appcompat.widget.AppCompatEditText
+                    android:id="@+id/edtUniversalBridgeMessage"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginBottom="@dimen/small_4"
+                    android:hint="@string/hint_universal_bridge_message"
+                    android:imeOptions="actionDone"
+                    android:inputType="text" />
```
Comment (khairul-alam-licon) [inline testapp/src/main/res/layout/qa_settings_activity.xml:273]: We can change to `textMultiLine` to input a json like this:

Screen Shot 2022-12-02 at 12 36 30

```suggestion
                    android:inputType="textMultiLine" />
```

File: testapp/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -122,4 +124,9 @@
     <string name="deeplink_activity_not_saved">No deeplink has been saved yet.</string>
     <string name="miniapp_by_url">MiniApp By URL</string>
 
+    <string name="universal_bridge">Universal Bridge</string>
+    <string name="hint_universal_bridge_message">Universal Bridge Message</string>
+    <string name="action_send_to_miniapp">Send to MiniApp</string>
+    <string name="universal_message_bridge_has_been_sent">Universal Bridge has been sent</string>
```
Comment (khairul-alam-licon) [inline testapp/src/main/res/values/strings.xml:130]:
```suggestion
    <string name="universal_message_send_to_miniapp">The Json info has been sent successfully to MiniApp via universal bridge.</string>
```

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt [hunk 12]
```diff
@@ -163,6 +164,17 @@ open class MiniAppMessageBridge {
...
 
+    /**
+     * handle Universal Bridge that represented as a json  from MiniApp.
+     **/
+    open fun sendJsonToHostApp(
+        jsonStr: String,
+        onSuccess: (jsonStr: String) -> Unit,
+        onError: (message: String) -> Unit
+    ) {
+        throw JsonInfoNotImplementedException()
```
Comment (munir-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppMessageBridge.kt:175]: @shohiebsense you can use this common implementation for consistancy
```
        throw MiniAppSdkException(ErrorBridgeMessage.NO_IMPL)
```
we generally throw this exception if any funtion is not overridden. I dont think there is a need to add a new one for this.

**Review by khairul-alam-licon — DISMISSED** [review]
Looks good 🥇 
Thank you for updating JS bridge and adding the sample app changes also. 
I have added few optional suggestions but, I think it's already ready to merge.
Thank you 🙇‍♂️

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,6 +1,10 @@
 ## CHANGELOG
 
 ### 5.1.0 (xxxx-xx-xx)
+**SDK**
+- **Feature:** Added `MiniAppMessageBridge.sendJsonToHostApp` interface to provide `Universal Bridge` for receiving messages from a MiniApp to the HostApp.
+- **Feature:** Added `NativeEventType.MINIAPP_RECEIVE_JSON_INFO` to provide `Universal Bridge` for sending messages from the HostApp to MiniApps.
```
Comment (munir-rakuten) [inline CHANGELOG.md:6]: Please update the `USERGUIDE` as well.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/userdata/QASettingsActivity.kt (hunk 9 above)
Comment (shohiebsense) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/userdata/QASettingsActivity.kt:140]: nice catch, thanks

**Comment by shohiebsense** [comment]
adjustment has been made, thanks a lot @munir-rakuten @khairul-alam-licon

File: miniapp/USERGUIDE.md [hunk 14]
```diff
@@ -270,6 +270,7 @@ There are some methods have a default implementation but the host app can overri
 | requestCustomPermissions     | ✅       |
 | shareContent                 | ✅       |
 | getHostEnvironmentInfo       | ✅       |
+| sendjsonToHostApp            | 🚫       |
```
Comment (khairul-alam-licon) [inline miniapp/USERGUIDE.md:273]:
```suggestion
| sendJsonToHostApp            | 🚫       |
```

File: miniapp/USERGUIDE.md [hunk 15]
```diff
@@ -519,6 +534,12 @@ You can also choose to override the default functionality and instead share the
 **API Docs:** [MiniAppMessageBridge.getHostEnvironmentInfo](api/com.rakuten.tech.mobile.miniapp.js/-mini-app-message-bridge/get-host-environment.html)
 
 The default functionality will provide information using `HostEnvironmentInfo` object to Mini App. Also, Host App can send it's environment information by implementing this function.
+    
+### Send Json To Host App
+**API Docs:** [MiniAppMessageBridge.sendJsonToHostApp](api/com.rakuten.tech.mobile.miniapp.js/-mini-app-message-bridge/send-json-to-host-app.html)
+
+The MiniApp is able to send the Universal Bridge. It can be either a message or an event for any purpose according to your need in `json` format.
```
Comment (munir-rakuten) [inline miniapp/USERGUIDE.md:541]: I think this `sendJsonToHostApp` only covers the sending a json string part, we can add event one in the event section
```suggestion
The MiniApp is able to send json format data as a string to Host App.
```

Comment (shohiebsense) [reply miniapp/USERGUIDE.md]: Refer to [this page
](https://confluence.rakuten-it.com/confluence/display/ONEAPP/Universal+Bridge)
`We need a generic interface(s) between Hostapp and MiniApp that can communicate freely. Interface(s) can be used for any purpose(By the Host app/MiniApp) according to their need.`

but no worries, change has been made.

**Comment by shohiebsense** [comment]
made some changes, reviewing it before 13:00 would be very appreciated.

**Comment by shohiebsense** [comment]
flow adjustments has been made

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppViewImpl.kt [hunk 16]
```diff
@@ -65,4 +66,15 @@ internal class MiniAppViewImpl(
...
     }
+
+    override fun sendJsonToMiniApp(message: String) {
+        message.let {
+            if (it.isNotBlank()) {
+                miniAppParameters.config.miniAppMessageBridge.dispatchNativeEvent(
+                    NativeEventType.MINIAPP_RECEIVE_JSON_INFO,
+                    message
+                )
+            }
```
Comment (rleojoseph) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppViewImpl.kt:77]: What if it is a blank ? We should return some error to Host app in that case as well right ?

File: miniapp/src/test/java/com/rakuten/tech/mobile/miniapp/js/UnimplementedMessageBridgeSpec.kt [hunk 17]
```diff
@@ -6,25 +6,31 @@ import org.mockito.kotlin.mock
 
 class UnimplementedMessageBridgeSpec : BridgeCommon() {
 
-    private val unImplementedMessageBrdige = object : MiniAppMessageBridge() {}
+    private val unimplementedMessageBridge = object : MiniAppMessageBridge() {}
```
Comment (rleojoseph) [inline miniapp/src/test/java/com/rakuten/tech/mobile/miniapp/js/UnimplementedMessageBridgeSpec.kt:9]: IDK why this change is required BTW ?

File: testapp/src/main/res/values/strings.xml [hunk 18]
```diff
@@ -105,6 +105,8 @@
     <string name="error_desc_miniapp_version_mismatch">Please check with the project \nadministrator. See our Help Center for \nmore details.</string>
     <string name="error_desc_promotional_text_missing">Promotional text is not available</string>
     <string name="error_desc_miniapp_too_many_request">"Mini app is requested too many times.</string>
+    <string name="error_send_json_to_host_app_invalid_string">Unable to send the message: please check the string and try again</string>
```
Comment (rleojoseph) [inline testapp/src/main/res/values/strings.xml:108]: Are we using this message ?

File: miniapp/src/test/java/com/rakuten/tech/mobile/miniapp/js/UnimplementedMessageBridgeSpec.kt (hunk 17 above)
Comment (shohiebsense) [reply miniapp/src/test/java/com/rakuten/tech/mobile/miniapp/js/UnimplementedMessageBridgeSpec.kt:9]: this is for the test case.

File: testapp/src/main/res/values/strings.xml (hunk 18 above)
Comment (shohiebsense) [reply testapp/src/main/res/values/strings.xml:108]: thanks for the heads up! It was a leftover.

**Comment by shohiebsense** [comment]
Thanks @rleojoseph, the adjustment has been made

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppViewImpl.kt (hunk 16 above)
Comment (shohiebsense) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppViewImpl.kt]: thanks, either the js SDK side will take this validation or.. if it is in here, I made the adjustment then.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt [hunk 19]
```diff
@@ -20,6 +22,7 @@ import kotlinx.android.synthetic.main.mini_app_main_layout.*
 class DemoAppMainActivity : BaseActivity() {
     private var currentNavController: LiveData<NavController>? = null
     private lateinit var binding: MiniAppMainLayoutBinding
+    val miniAppIdAndViewMap = hashMapOf<String, MiniAppView>()
```
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt:25]: if we use same project id and subscription key in both List-I and List-II in demo app, then we are supposed to get multiple miniapps with same `MiniAppId`, therefore we should store the duplicates too. Otherwise, it will work on only one MiniApp which was opened firstly as we are eliminating the duplicate `MiniAppId` using this `HashMap`. And, the other MiniApps with same `MiniAppId` will not work.

I think, `HashMap.put` will be inconvenient to store multiple `MiniAppView`s with duplicate `MiniAppId`. In that case, we can apply other way i.e. `val miniAppIdAndViewMap = mutableListOf<Pair<String, MiniAppView>>()`

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt [hunk 20]
```diff
@@ -147,4 +153,21 @@ class DemoAppMainActivity : BaseActivity() {
             fragment
         } else null
     }
+
+    override fun onStart() {
+        super.onStart()
+        miniAppIdAndViewMap.forEach {
+            with(AppSettings.instance.universalBridgeState) {
```
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt:160]: I'm not sure, if there is any spec to send the Json message to MiniApp from anywhere from HostApp or only from the QA settings page 🤔 cc: @rleojoseph @munir-rakuten @GondiTilak 

#### If only from the QA settings page:
Then - we don't need to cache anything like `UniversalBridgeState` or even the message content if we can send the message directly from the `QASettingsActivity` after pressing the button. Inside, `QASettingsActivity` we can access `miniAppIdAndViewMap` which was defined in `DemoAppMainActivity`.

Then, we don't need to do `sendJsonToMiniApp` at `onStart` here...

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppViewImpl.kt [hunk 21]
```diff
@@ -65,4 +66,17 @@ internal class MiniAppViewImpl(
...
         }
     }
+
+    override fun sendJsonToMiniApp(message: String, onFailed: () -> Unit) {
+        message.let {
+            if (it.isNotBlank()) {
+                miniAppParameters.config.miniAppMessageBridge.dispatchNativeEvent(
+                    NativeEventType.MINIAPP_RECEIVE_JSON_INFO,
+                    message
+                )
```
Comment (munir-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppViewImpl.kt:76]:
```suggestion
when (miniAppParameters) {
                    is MiniAppParameters.DefaultParams -> (miniAppParameters as MiniAppParameters.DefaultParams).config.miniAppMessageBridge.dispatchNativeEvent(
                        NativeEventType.MINIAPP_RECEIVE_JSON_INFO,
                        message
                    )
                    is MiniAppParameters.InfoParams -> (miniAppParameters as MiniAppParameters.InfoParams).config.miniAppMessageBridge.dispatchNativeEvent(
                        NativeEventType.MINIAPP_RECEIVE_JSON_INFO,
                        message
                    )
                    is MiniAppParameters.UrlParams -> (miniAppParameters as MiniAppParameters.UrlParams).config.miniAppMessageBridge.dispatchNativeEvent(
                        NativeEventType.MINIAPP_RECEIVE_JSON_INFO,
                        message
                    )
                }
```

can you change it to these,  adding overridden parameter in the `sealed` class is not necessary now, i am a bit concerned that `config` might be effected the parameters inside it, but we can improve this later with some more tests. thank you.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppParameters.kt [hunk 22]
```diff
@@ -6,7 +6,9 @@ import com.rakuten.tech.mobile.miniapp.MiniAppInfo
 /**
  * This class can be used in the HostApp to create the mini app views using mini app id.
  */
-sealed class MiniAppParameters {
+sealed class MiniAppParameters(
+    open val config: MiniAppConfig,
```
Comment (munir-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppParameters.kt:10]:
```suggestion
```
can you change it to these,  adding overridden parameter in the `sealed` class is not necessary now, i am a bit concerned that `config` might be effected the parameters inside it and the miniapps itself, but we can improve this later with some more tests. thank you.

**Review by munir-rakuten — COMMENTED** [review]
Added some suggestions.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt (hunk 20 above)
Comment (munir-rakuten) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt:160]: I also think no need for any shared preference to cache the data, just send the `miniAppView` list to settings page and send the event the time user press the send  button.

**Comment by shohiebsense** [comment]
Thanks a lot for the comments, now look to them

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppParameters.kt (hunk 22 above)
Comment (shohiebsense) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppParameters.kt:10]: Thanks for the comment,  it was covered in the unit tests and it goes well. Would you explain what kind of effect that be?

do you mean to remove it back to the origin and suggest doing the same lines of code of finding which type of `MiniAppParameters` as in the `load()`?

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt (hunk 20 above)
Comment (shohiebsense) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt:160]: I don't get the point here.. Considering `DemoAppMainActivity` and `QaSettingsActivity` are having different lifecycles, would you explain a bit about why we don't need these...? 🤔

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppView.kt [hunk 23]
```diff
@@ -63,4 +63,11 @@ abstract class MiniAppView internal constructor() {
         fromCache: Boolean = false,
         onComplete: (MiniAppDisplay?, MiniAppSdkException?) -> Unit
     )
+
+    /**
+     * send a generic message
+     * to MiniApp using [com.rakuten.tech.mobile.miniapp.js.NativeEventType.MINIAPP_RECEIVE_JSON_INFO].
+     * @param message the content that will send to the MiniApp
+     */
+    abstract fun sendJsonToMiniApp(message: String)
```
[the remark this answers is not in the document]
Comment (shohiebsense) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppView.kt]: ok that will do

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt (hunk 20 above)
Comment (shohiebsense) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt:160]: aaah... Bringing the HashMap. Alright

Comment (shohiebsense) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt:160]: let's try using static then

Comment (khairul-alam-licon) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/DemoAppMainActivity.kt:160]: I think, we can declare it as a global variable e.g. 

```
val miniAppIdAndViewMap = ...
class DemoAppMainActivity : BaseActivity() { ... }
```

**Comment by shohiebsense** [comment]
The adjustment has been made, please check again.

**Comment by shohiebsense** [comment]
adjustment made, may have a look

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/fragments/MiniAppDisplayFragment.kt [hunk 24]
```diff
@@ -104,20 +104,30 @@ class MiniAppDisplayFragment : BaseFragment(), PreloadMiniAppWindow.PreloadMiniA
...
+            } else {
                 initializeMiniAppDisplay(activity)
+            }
         }
-        return binding.root
     }
 
     @Suppress("LongMethod", "ComplexMethod")
-    private fun initializeMiniAppDisplay(activity: Activity) {
+    private fun initializeMiniAppDisplay(activity: DemoAppMainActivity) {
```
Comment (munir-rakuten) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/miniapptabs/fragments/MiniAppDisplayFragment.kt:130]: is that necessary to cast it as `DemoAppMainActivity`? now this fragment can not be work with out `DemoAppMainActivity`, we can not use this fragment anywhere, we might use this fragment in the future so i think it's better not to cast to a specific activity. just use
```
if(activity is DemoAppMainActivity)
   (activity as DemoAppMainActivity).somefunction()

```

**Review by munir-rakuten — APPROVED** [review]
Looks good to me, added just one suggestion, good work👍

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppParameters.kt (hunk 22 above)
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/view/MiniAppParameters.kt]: yes @shohiebsense

**Review by khairul-alam-licon — APPROVED** [review]
Looks good 🎉 

Also, please add `ActionType.JSON_INFO` into `MessageBridgeRatDispatcher` to send it to the analytics.
